internal/mode/static/state/graph/route_common.go

```diff
@@ -0,0 +1,567 @@
+type ParentRef struct {
+	// Attachment is the attachment status of the ParentRef. It could be nil. In that case, NGF didn't attempt to
+	// attach because of problems with the Route.
+	Attachment *ParentRefAttachmentStatus
+	// Gateway is the NamespacedName of the referenced Gateway
+	Gateway types.NamespacedName
+	// Idx is the index of the corresponding ParentReference in the Route.
+	Idx int
+}
+
+// ParentRefAttachmentStatus describes the attachment status of a ParentRef.
+type ParentRefAttachmentStatus struct {
+	// AcceptedHostnames is an intersection between the hostnames supported by an attached Listener
+	// and the hostnames from this Route. Key is listener name, value is list of hostnames.
+	AcceptedHostnames map[string][]string
+	// FailedCondition is the condition that describes why the ParentRef is not attached to the Gateway. It is set
+	// when Attached is false.
+	FailedCondition conditions.Condition
+	// Attached indicates if the ParentRef is attached to the Gateway.
+	Attached bool
+}
+
+type RouteType string
+
+const (
+	// RouteTypeHTTP indicates that the RouteType of the L7Route is HTTP
+	RouteTypeHTTP RouteType = "http"
+	// RouteTypeGRPC indicates that the RouteType of the L7Route is gRPC
+	RouteTypeGRPC RouteType = "grpc"
+)
+
+// RouteKey is the unique identifier for a L7Route
+type RouteKey struct {
+	NamespacedName types.NamespacedName
+	RouteType      RouteType
+}
+
+// L7Route is the generic type for the layer 7 routes, HTTPRoute and GRPCRoute
+type L7Route struct {
+	// Source is the source Gateway API object of the Route.
+	Source client.Object
+	// RouteType is the type (http or grpc) of the Route.
+	RouteType RouteType
+	// Spec is the L7RouteSpec of the Route
+	Spec L7RouteSpec
+	// ParentRefs describe the references to the parents in a Route.
+	ParentRefs []ParentRef
+	// SrcParentRefs contain the Gateway API references to the parents in a Route.
+	SrcParentRefs []v1.ParentReference
+	// Conditions define the conditions to be reported in the status of the Route.
+	Conditions []conditions.Condition
+	// Valid indicates if the Route is valid.
+	Valid bool
+	// Attachable indicates if the Route is attachable to any Listener.
+	Attachable bool
+}
+
+type L7RouteSpec struct {
+	// Hostnames defines a set of hostnames used to select a Route used to process the request.
+	Hostnames []v1.Hostname
+	// Rules are the list of HTTP matchers, filters and actions.
+	Rules []RouteRule
+}
+
+type RouteRule struct {
+	// Matches define the predicate used to match requests to a given action.
+	Matches []v1.HTTPRouteMatch
+	// Filters define processing steps that must be completed during the request or response lifecycle.
+	Filters []v1.HTTPRouteFilter
+	// RouteBackendRefs are a wrapper for v1.BackendRef and any BackendRef filters from the HTTPRoute or GRPCRoute.
+	RouteBackendRefs []RouteBackendRef
+	// BackendRefs is an internal representation of a backendRef in a Route.
+	BackendRefs []BackendRef
```

Comment:
Why do we need both RouteBackendRefs and BackendRefs?

Reply:
RouteBackendRefs are the original, pre-validated BackendRefs from the Route. BackendRefs are the validated and processed refs. We need both because we do the backend refs validation separately, so we've built the L7Route by that point before we process the BackendRefs.